pinot-core/src/main/java/org/apache/pinot/core/query/aggregation/function/customobject/DistinctTable.java

```diff
@@ -0,0 +1,297 @@
+        orderByAsc[i] = selectionSort.isIsAsc();
+      }
+      _sortedRecords = new PriorityQueue<>(initialCapacity, (record1, record2) -> {
+        Object[] values1 = record1.getValues();
+        Object[] values2 = record2.getValues();
+        for (int i = 0; i < numOrderByColumns; i++) {
+          Comparable valueToCompare1 = (Comparable) values1[orderByColumnIndexes[i]];
+          Comparable valueToCompare2 = (Comparable) values2[orderByColumnIndexes[i]];
+          int result =
+              orderByAsc[i] ? valueToCompare2.compareTo(valueToCompare1) : valueToCompare1.compareTo(valueToCompare2);
+          if (result != 0) {
+            return result;
+          }
+        }
+        return 0;
+      });
+    } else {
+      _sortedRecords = null;
+    }
+    _records = null;
+  }
+
+  /**
+   * Returns the {@code DataSchema} of the {@code DistinctTable}.
+   */
+  public DataSchema getDataSchema() {
+    return _dataSchema;
+  }
+
+  /**
+   * Returns the number of unique records within the {@code DistinctTable}.
+   */
+  public int size() {
+    if (_uniqueRecords != null) {
+      // Server-side
+      return _uniqueRecords.size();
+    } else {
+      // Broker-side
+      return _records.size();
+    }
+  }
+
+  /**
+   * Adds a record into the DistinctTable and returns whether more records should be added. No more records should be
+   * added iff there is no order-by column and enough unique records have been collected.
+   */
+  public boolean add(Record record) {
+    if (_uniqueRecords.contains(record)) {
+      return true;
+    }
+    if (_sortedRecords != null) {
+      if (_sortedRecords.size() < _limit) {
+        _uniqueRecords.add(record);
+        _sortedRecords.offer(record);
+      } else {
+        Record leastRecord = _sortedRecords.peek();
+        if (_sortedRecords.comparator().compare(record, leastRecord) > 0) {
+          _uniqueRecords.remove(leastRecord);
+          _uniqueRecords.add(record);
+          _sortedRecords.poll();
+          _sortedRecords.offer(record);
+        }
+      }
+      return true;
+    } else {
+      _uniqueRecords.add(record);
+      return _uniqueRecords.size() < _limit;
+    }
+  }
+
+  /**
+   * Returns {@code true} if no more records should be added, {@code false otherwise}. No more records should be added
+   * iff there is no order-by columns and enough unique records have been collected.
+   */
+  public boolean shouldNotAddMore() {
+    return _sortedRecords == null && _uniqueRecords.size() == _limit;
+  }
+
+  /*
+   * SERVER ONLY METHODS
+   */
+
+  /**
+   * (Server-side) Merges another {@code DistinctTable} into the current one.
+   * <p>NOTE: {@code DistinctTable} on Server-side has non-null {@code _uniqueRecords}.
+   */
+  public void serverSideMerge(DistinctTable distinctTable) {
+    if (shouldNotAddMore()) {
+      return;
+    }
+    for (Record record : distinctTable._uniqueRecords) {
+      if (!add(record)) {
+        return;
+      }
```

Comment:
Here we seem to be doing extra check on the return value of add() for each record. If there is order by, add() will never return false right?

Reply:
Yes, re-structured the code to avoid that